pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -215,130 +208,132 @@ func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
-		if apierrors.IsNotFound(err) {
-			klog.V(5).InfoS(fmt.Sprintf("Something concurrently deleted the %s", wrapper.TypeName()), "name", name)
-			return nil
+	wantedNames := namesOfBootstrapObjects(boots)
+	for _, object := range current {
+		name := object.GetName()
+		if wantedNames.Has(name) {
+			continue
 		}
-
-		return fmt.Errorf("failed to delete the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
-	}
-
-	klog.V(2).InfoS(fmt.Sprintf("Successfully deleted the %s", wrapper.TypeName()), "name", name)
-	return nil
-}
-
-// getDanglingBootstrapObjectNames returns a list of names of bootstrap
-// configuration objects that are potentially candidates for deletion from
-// the cluster, given a set of bootstrap and current configuration.
-//   - bootstrap: a set of hard coded configuration kube-apiserver maintains in-memory.
-//   - current: a set of configuration objects that exist on the cluster
-//
-// Any object present in current is added to the list if both a and b are true:
-//
-//	a. the object in current is missing from the bootstrap configuration
-//	b. the object has the designated auto-update annotation key
-//
-// This function shares the common logic for both FlowSchema and
-// PriorityLevelConfiguration type and hence it accepts metav1.Object only.
-func getDanglingBootstrapObjectNames(bootstrap sets.String, current []metav1.Object) []string {
-	if len(current) == 0 {
-		return nil
-	}
-
-	candidates := make([]string, 0)
-	for i := range current {
-		object := current[i]
-		if _, ok := object.GetAnnotations()[flowcontrolv1beta3.AutoUpdateAnnotationKey]; !ok {
+		var value string
+		var ok, autoUpdate bool
+		var err error
+		if value, ok = object.GetAnnotations()[flowcontrolv1beta3.AutoUpdateAnnotationKey]; !ok {
 			// the configuration object does not have the annotation key,
 			// it's probably a user defined configuration object,
 			// so we can skip it.
+			klog.V(5).InfoS("Skipping deletion of APF object with no "+flowcontrolv1beta3.AutoUpdateAnnotationKey+" annotation", "name", name)
 			continue
 		}
-
-		if _, ok := bootstrap[object.GetName()]; !ok {
-			candidates = append(candidates, object.GetName())
+		autoUpdate, err = strconv.ParseBool(value)
+		if err != nil {
+			// Log this because it is not an expected situation.
+			klog.V(4).InfoS("Skipping deletion of APF object with malformed "+flowcontrolv1beta3.AutoUpdateAnnotationKey+" annotation", "name", name, "annotationValue", value, "parseError", err)
+			continue
+		}
+		if !autoUpdate {
+			klog.V(5).InfoS("Skipping deletion of APF object with "+flowcontrolv1beta3.AutoUpdateAnnotationKey+"=false annotation", "name", name)
+			continue
+		}
+		// TODO: expectedResourceVersion := object.GetResourceVersion()
+		err = object.delete(ctx /* TODO: expectedResourceVersion */)
+		if err == nil {
+			klog.V(2).InfoS(fmt.Sprintf("Successfully deleted the unwanted %s", boots.typeName()), "name", name)
+			continue
 		}
+		if apierrors.IsConflict(err) {
```

Comment:
if i am not mistaken, this seems unrelated to the refactor, and the context wiring also not related to the refactor, if not a separate PR, can we have three separate commits in this pr? thanks!